### DISCUSSION TITLE
Runtime environment variables issue (regarding Procfile)

### DISCUSSION THREAD

**Discussion by Rvarman0110** [paketo-buildpacks/feedback#252] [Java Team]
Hello everyone, I am currently facing an issue as my image has this condition to use this environment variable during the runtime 

**-Dspring.config.additional-location=file:./data/labels.properties -XX:InitialRAMPercentage=35.0 -XX:MaxRAMPercentage=70.0 -XX:+PrintGCDetails -Xloggc:/tmp/gc-${HOSTNAME}-%t.log -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=/tmp/dump-${HOSTNAME}-%t.hprof**

As I am using paketo buildpacks to build this image I am getting a predefined process type which has the command to execute the Jar 

How can I add the environment variables that should work on runtime of that image ?
I read about adding the runtime env in project.toml file but I didnt get any clear idea on this , is there any example to show how  I can achieve this runtime env in project.toml file, It would be very helpful for me.

I also took a look on Procfile , say I am running java command in that procfile **Do I need java to be installed on my run stack image to execute the java commands which is inside my procfile** ?

Thanks

**Comment by dmikusa** [comment]
Environment variables set when you run `pack` are by default **only** set at build time. Using `project.toml` is no different, they are build time only as well. If you want to set env variables at runtime, then you need to tell your container orchestrator (ie. Docker, K8s, etc...) to set them. For example, `docker run -e FOO=bar ...`. 

The other option is to use the environment variables buildpack and embed the env variables in the image. That can be done [by following the instructions here](https://paketo.io/docs/howto/configuration/#image-embedded-environment-variables).

As far as Procfile. We don't really recommend using that with the Java buildpacks. You can nothing will stop you, but it adds unnecessary complication and can easily break things. Instead, what I would suggest is that you set flags to the JVM by setting `JAVA_TOOL_OPTIONS=<your flags>` at **runtime**. The JVM will read this env variable and apply your flags. You should be able to copy all of your settings above into that env variable.

Two other notes:

1. Be careful with `-XX:InitialRAMPercentage=35.0 -XX:MaxRAMPercentage=70.0`. The buildpack already configures memory usage based on the container's limits. This happens dynamically and without users needing to do anything extra. Setting multiple memory settings, best case, just makes things confusing. 

2. I would also point you to `BPL_HEAP_DUMP_PATH` which is functionality the buildpack exposes. If you set this to a path at runtime, it will automatically add the heap dump related settings. It is specified at runtime so you can turn that functionality on/off without rebuilding. See [the docs here](https://github.com/paketo-buildpacks/bellsoft-liberica?tab=readme-ov-file#configuration) for this setting and more.

**Reply by Rvarman0110** [reply]
Thank you for the clarification, I will try these methods
